src/main/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProvider.java

```diff
@@ -0,0 +1,230 @@
+  }
+
+  public static InputStream decryptStream(
+      InputStream inputStream,
+      String keyBase64,
+      String cekIvBase64,
+      String kekIvBase64,
+      RemoteStoreFileEncryptionMaterial encMat,
+      byte[] cekAad,
+      byte[] kekAad)
+      throws InvalidKeyException, BadPaddingException, IllegalBlockSizeException,
+          InvalidAlgorithmParameterException, NoSuchPaddingException, NoSuchAlgorithmException {
+    byte[] keyBytes = Base64.getDecoder().decode(keyBase64);
+    byte[] ivBytes = Base64.getDecoder().decode(cekIvBase64);
+    byte[] kekIvBytes = Base64.getDecoder().decode(kekIvBase64);
+    byte[] decodedKey = Base64.getDecoder().decode(encMat.getQueryStageMasterKey());
+
+    byte[] fileKeyBytes = decryptKey(decodedKey, kekIvBytes, keyBytes, kekAad);
+    return decryptContentFromStream(inputStream, ivBytes, fileKeyBytes, decodedKey, cekAad);
+  }
+
+  private static CipherInputStream decryptContentFromStream(
+      InputStream inputStream, byte[] ivBytes, byte[] fileKeyBytes, byte[] decodedKey, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, NoSuchPaddingException,
+          NoSuchAlgorithmException {
+    GCMParameterSpec cekGcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, ivBytes);
+    SecretKey fileKey = new SecretKeySpec(fileKeyBytes, 0, decodedKey.length, AES);
+    Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+    fileCipher.init(Cipher.DECRYPT_MODE, fileKey, cekGcmParameterSpec);
+    if (aad != null) {
+      fileCipher.updateAAD(aad);
+    }
+    return new CipherInputStream(inputStream, fileCipher);
+  }
+
+  public static void decryptFile(
+      File file,
+      String keyBase64,
+      String ivBase64,
+      String kekIvBase64,
+      RemoteStoreFileEncryptionMaterial encMat,
+      byte[] cekAad,
+      byte[] kekAad)
+      throws InvalidKeyException, IllegalBlockSizeException, BadPaddingException,
+          InvalidAlgorithmParameterException, IOException, NoSuchPaddingException,
+          NoSuchAlgorithmException {
+    byte[] keyBytes = Base64.getDecoder().decode(keyBase64);
+    byte[] cekIvBytes = Base64.getDecoder().decode(ivBase64);
+    byte[] kekIvBytes = Base64.getDecoder().decode(kekIvBase64);
+    byte[] decodedKey = Base64.getDecoder().decode(encMat.getQueryStageMasterKey());
+
+    byte[] fileKeyBytes = decryptKey(decodedKey, kekIvBytes, keyBytes, kekAad);
+    decryptContentFromFile(file, fileKeyBytes, cekIvBytes, cekAad);
+  }
+
+  private static void decryptContentFromFile(
+      File file, byte[] fileKeyBytes, byte[] cekIvBytes, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, IOException,
+          NoSuchPaddingException, NoSuchAlgorithmException {
+    final SecretKey fileKey = new SecretKeySpec(fileKeyBytes, AES);
+    GCMParameterSpec cekGcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, cekIvBytes);
```

Comment:
why do we have some fields final some not?

Reply:
I removed the final keyword.